Blockchain/task/src/blockchain/management/BlockChain.java

```diff
@@ -0,0 +1,129 @@
+package blockchain.management;
+
+import java.io.*;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class BlockChain implements Iterable<Block>, Serializable {
+    private Block head;
+    private Block tail;
+    private long size;
+    private AtomicLong lastId = new AtomicLong(0);
+    private AtomicInteger N = new AtomicInteger(0);
+    private NavigableSet<Transaction> transactions = new TreeSet<>(Comparator.comparingLong(Transaction::getId));
+    private ArrayList<MoneyHandler> moneyHandlers = new ArrayList<>();
+
+    public BlockChain() {}
+
+    public void checkValidityOf(Block beingChecked) throws InvalidObjectException {
+        if(beingChecked.isProved(N.get()) && (tail == null
+                || Objects.equals(tail.hash(), beingChecked.getPreviousHash()))) return;
+
+        throw new InvalidObjectException("Block isn't valid");
+    }
+
+    public byte checkN(long generationTime) {
+        if(generationTime > 1_000_000_000) {
+            N.decrementAndGet();
+            return -1;
+        }
+
+        if(generationTime < 10_000_000) {
+            N.incrementAndGet();
+            return 1;
+        }
+
+        return 0;
+    }
+
+    public void printBlocks(int bound) {
+        if(0 > bound || bound > size)
+            throw new IllegalArgumentException();
+
+        int counter = 0;
+        for (Block block: this) {
+            block.print();
+            if(counter++ == bound) return;
+        }
+    }
+
+    public boolean addBlock(Block newBlock) {
+        try {
+            checkValidityOf(newBlock);
+        } catch (InvalidObjectException e) {
+            return false;
+        }
+
+        if(head == null) head = newBlock;
+        else tail.next = newBlock;
```

Comment:
do you really need this line of code? tail.next reference will be erased by the next line?